lib/compat/wordpress-6.3/get-global-styles-and-settings.php

```diff
@@ -0,0 +1,188 @@
+		// feature selectors later on.
+		$root_selector = null;
+
+		if ( $use_editor_selectors && isset( $block_type->editor_selectors['root'] ) ) {
+			// Prefer editor selectors if specified.
+			$root_selector = $block_type->editor_selectors['root'];
+		} elseif ( $has_selectors && isset( $block_type->selectors['root'] ) ) {
+			// Use the selectors API if available.
+			$root_selector = $block_type->selectors['root'];
+		} elseif ( isset( $block_type->supports['__experimentalSelector'] ) && is_string( $block_type->supports['__experimentalSelector'] ) ) {
+			// Use the old experimental selector supports property if set.
+			$root_selector = $block_type->supports['__experimentalSelector'];
+		} else {
+			// If no root selector found, generate default block class selector.
+			$block_name    = str_replace( '/', '-', str_replace( 'core/', '', $block_type->name ) );
+			$root_selector = ".wp-block-{$block_name}";
+		}
+
+		// Return selector if it's the root target we are looking for.
+		if ( 'root' === $target ) {
+			return $root_selector;
+		}
+
+		// If target is not `root` or `duotone` we have a feature or subfeature
+		// as the target. If the target is a string convert to an array.
+		if ( is_string( $target ) ) {
+			$target = explode( '.', $target );
+		}
+
+		// Feature Selectors ( May fallback to root selector ).
+		if ( 1 === count( $target ) ) {
+			$fallback_selector = $fallback ? $root_selector : null;
+
+			// Look for selector under `feature.root`.
+			$path = array_merge( $target, array( 'root' ) );
+
+			// Use editor specific selector if available.
+			if ( $use_editor_selectors ) {
+				$feature_selector = _wp_array_get( $block_type->editor_selectors, $path, null );
+
+				if ( $feature_selector ) {
+					return $feature_selector;
+				}
+
+				// Check if feature selector set via shorthand.
+				$feature_selector = _wp_array_get( $block_type->editor_selectors, $target, null );
+
+				// Only return if a selector was found.
+				if ( is_string( $feature_selector ) ) {
+					return $feature_selector;
+				}
+			}
+
+			// Prefer the selectors API if available.
+			if ( $has_selectors ) {
+				$feature_selector = _wp_array_get( $block_type->selectors, $path, null );
+
+				if ( $feature_selector ) {
+					return $feature_selector;
+				}
+
+				// Check if feature selector set via shorthand.
+				$feature_selector = _wp_array_get( $block_type->selectors, $target, null );
+
+				return is_string( $feature_selector ) ? $feature_selector : $fallback_selector;
+			}
+
+			// Try getting old experimental supports selector value.
+			$path             = array_merge( $target, array( '__experimentalSelector' ) );
+			$feature_selector = _wp_array_get( $block_type->supports, $path, null );
+
+			// Nothing to work with, provide fallback or null.
+			if ( null === $feature_selector ) {
+				return $fallback_selector;
+			}
+
+			// Scope the feature selector by the block's root selector.
```

Comment:
Documenting what I'm learning. No need to respond if it's clear to everyone else.

In the experimental API, feature-level selectors would be wrapped with the root selector, but in the stable one we need to provide the full selector for each level. Using the calendar block as an example. The resulting block selector is `.wp-block-calendar table, .wp-block-calendar th` and

- the old experimental API declares it as:

```json
{
"supports": {
  "color": {
    "__experimentalSelector": "table, th"
  }
}
}
```

- the new stable API declares it as:

```json
{
  "selectors": {
    "color": ".wp-block-calendar table, .wp-block-calendar th"
  }
}
```

The new stable API is more explicit (the reader understand the output immediately) and it's more flexible (allows for use cases where the feature-level selector doesn't need to contain the block wrapper class).

I wanted to bring this anyway in case we did it that way in the old experimental API for a particular use case we're missing it now.

Reply:
I thought only `__experimentalDuotone` had the additional wrapper added previously. I added it that way so you couldn't accidentally select something outside of the block and have duotone filters applied to all `img` on the page, for example.

Reply:
> I thought only __experimentalDuotone had the additional wrapper added previously.

No, feature-level selectors got [scoped by the block's root selector](https://github.com/WordPress/gutenberg/pull/46496/files#diff-b03597cc3da199e5aa5a94950e8241135904853e7c3b82d758e42744878afae7L880-L883) as well.

I understand the point regarding preventing a selector that incorrectly applies the duotone to all images. Do we really need to protect against block authors adding an incorrect selector to the block.json though?

Having the explicit selectors without automatic scoping by the block's root selector also allows for chained selectors and continued targeting of the block's wrapper e.g. `.wp-block-name:not(.is-style-ignored)`.

Reply:
> Do we really need to protect against block authors adding an incorrect selector to the block.json though?

No, I was just trying to provide the reasoning for why duotone was done that way.

> Having the explicit selectors without automatic scoping by the block's root selector also allows for chained selectors and continued targeting of the block's wrapper e.g. `.wp-block-name:not(.is-style-ignored)`.

The other way to support that would be SASS-style `&:not(.is-style-ignored)` which is what I originally started duotone to support. I ended up not going that way because it would have made things more complicated, and I didn't really need the `&` support anyway.

---

So, to be clear, I am okay with just typing out the whole selector.